src/ledger/database/hashmap.zig

```diff
@@ -178,9 +180,17 @@ pub fn SharedHashMapDB(comptime column_families: []const ColumnFamily) type {
                         const cf_index, const key = delete_ix;
                         self.maps[cf_index].delete(self.allocator, key);
                     },
-                    .delete_range => {
-                        // TODO: also add to database tests
-                        @panic("not implemented");
+                    .delete_range => |delete_range_ix| {
+                        const cf_index, const start, const end = delete_range_ix;
+                        const keys, _ = self.maps[cf_index].map.range(start, end);
+                        const to_delete = try batch.allocator.alloc([]const u8, keys.len);
+                        defer batch.allocator.free(to_delete);
+                        for (keys, 0..) |key, i| {
```

Comment:
Not sure I understand the need of first filling the `to_delete` array first, to then iterate over it, retrieving the keys and using that as the `delete_key` in

```
self.maps[cf_index].delete(self.allocator, delete_key);
```

Any reason for not directly using the key from `keys` as the `delete_key`?

Reply:
Good catch. I was wondering if someone would comment on this.

The problem is that it's not safe to mutate the hash map while iterating over a pointer to its data. This loop is iterating over the data returned by `ArrayHashMap.keys`.

```zig
/// Returns the backing array of keys in this map. Modifying the map may
/// invalidate this array. Modifying this array in a way that changes
/// key hashes or key equality puts the map into an unusable state until
/// `reIndex` is called.
pub fn keys(self: Self) []K {
    return self.unmanaged.keys();
}
```